generator/internal/dart/darttemplate.go

```diff
@@ -357,19 +355,165 @@ func annotateMessage(m *api.Message, state *api.APIState, packageMapping map[str
 		constructorBody = " {\n    _validate();\n  }"
 	}
 
+	_, hasCustomEncoding := usesCustomEncoding[m.ID]
 	toStringLines := createToStringLines(m)
 
 	m.Codec = &messageAnnotation{
-		Name:            messageName(m),
-		DocLines:        formatDocComments(m.Documentation, state),
-		ConstructorBody: constructorBody,
-		BasicFields: language.FilterSlice(m.Fields, func(s *api.Field) bool {
-			return !s.IsOneOf
-		}),
-		HasFields:        len(m.Fields) > 0,
-		HasToStringLines: len(toStringLines) > 0,
-		ToStringLines:    toStringLines,
+		Name:              messageName(m),
+		DocLines:          formatDocComments(m.Documentation, state),
+		ConstructorBody:   constructorBody,
+		HasFields:         len(m.Fields) > 0,
+		HasCustomEncoding: hasCustomEncoding,
+		FromJsonLines:     createFromJsonLines(m, state),
+		ToJsonLines:       createToJsonLines(m, state),
+		HasToStringLines:  len(toStringLines) > 0,
+		ToStringLines:     toStringLines,
+	}
+}
+
+func createFromJsonLines(message *api.Message, state *api.APIState) []string {
+	_, hasCustomEncoding := usesCustomEncoding[message.ID]
+	if hasCustomEncoding {
+		return []string{}
+	}
+
+	lines := []string{}
+
+	for _, field := range message.Fields {
+		codec := field.Codec.(*fieldAnnotation)
+		name := codec.Name
+		message := state.MessageByID[field.TypezID]
+		typeName := ""
+
+		isCustom := false
+		isList := field.Repeated
+		isMessage := field.Typez == api.MESSAGE_TYPE
+		isEnum := field.Typez == api.ENUM_TYPE
+		isMap := message != nil && message.IsMap
+		isMessageMap := isMap && message.Fields[1].Typez == api.MESSAGE_TYPE
+		isRequired := codec.Required
+
+		if isMessage {
+			typeName = messageName(message)
+			_, isCustom = usesCustomEncoding[message.ID]
+		} else if isEnum {
+			enum := state.EnumByID[field.TypezID]
+			typeName = enumName(enum)
+		}
+
+		opt := ""
+		bang := "!"
+		if !isRequired {
+			opt = "?"
+			bang = ""
+		}
+
+		if isMessageMap {
+			// message maps
+			//   name: $toMap(json['name'], Status.fromJson)!,
+			//   name: $toMap(json['name'], Status.fromJson),
+			lines = append(lines, fmt.Sprintf("%s: $toMap(json['%s'], %s.fromJson)%s,", name, name, typeName, bang))
+		} else if isMap {
```

Comment:
I think I see what you are doing here: the treatment for the fields is very different depending on whether it is optional, repeated, a map, etc.  In Rust we did something different.  We partition the fields:

https://github.com/googleapis/google-cloud-rust/blob/f8d2b7e78da336b6ad1308060904092c335144bd/generator/internal/rust/annotate.go#L328-L334

and then use these partitions in the mustache template:

https://github.com/googleapis/google-cloud-rust/blob/f8d2b7e78da336b6ad1308060904092c335144bd/generator/internal/rust/templates/common/message.mustache#L70-L83

Do you think that may help here?

Reply:
Yeah, what we need to write out depends on whether something is singular, nullable, repeated, a map of primitives, a message w/ a custom encoding, and map of messages, ... This complex if statement was the simplest impl. I could get to.

I'll give some throught to whether we could partition things up front in a way that would let us push some of the logic down to the templates.